KK-1105 | fix: header menu localized routing

melniiv opened the pull request:
## Description
Fixes the localised pages routing issue. 
Note. Not all the pages have translations in test environment. If translated page is missing, user is redirected to the root of the website

<!-- Describe your changes in detail -->

## Context

<!-- Why is this change required? What problem does it solve? -->
<!-- Leave a link to the Jira ticket for posterity. -->

[KK-1105](https://helsinkisolutionoffice.atlassian.net/browse/KK-1105)

## How Has This Been Tested?

<!-- Explain what automated and manual testing approaches you have used -->

## Manual Testing Instructions for Reviewers

<!-- Make it easy for reviewers to test your changes by providing instructions -->

## Screenshots

<!-- Add screenshots if appropriate -->


[KK-1105]: https://helsinkisolutionoffice.atlassian.net/browse/KK-1105?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

terovirtanen commented:
**KUKKUU-UI branch is deployed to platta: https://kukkuu-ui-pr564.dev.hel.ninja** :rocket::rocket::rocket:

nikomakela commented:
From the ticket KK-1105, I tested these 2 steps while my local machine was connected to the production CMS:

✅ 1. Go to https://kummilapset.hel.fi/en/sponsors/ and change the language into Finnish from header link. When the page https://kummilapset.hel.fi/fi/sponsors/  opens, the content on this page should be in Finnish. It’s not. The user should be directed to page https://kummilapset.hel.fi/fi/kummitoimijat/ 

✅ 2. Go to [Kulttuurin kummilapset](https://kummilapset.hel.fi/sv/sv-kenelle/)  and change the language into English from header link. When https://kummilapset.hel.fi/en/sv-kenelle/  opens, the content on this page be in English. It’s not. The user should be directed to page https://kummilapset.hel.fi/en/sponsors/

❌ However, in the ticket comments there is also a mention about the "käyttöehdot" link in the footer. For some reason, that does not work. It could be because of CMS data is set wrong?

The URL becomes `/fi/terms`.
<img width="1205" alt="image" src="https://github.com/City-of-Helsinki/kukkuu-ui/assets/389204/01d2ee80-b073-4efb-a09a-9d2a5ea74e79">

The menu item is not indicated as active:
<img width="1225" alt="image" src="https://github.com/City-of-Helsinki/kukkuu-ui/assets/389204/4325e44b-da00-4d31-946f-529c4645a2e5">

nikomakela commented on src/domain/app/navigation/Navigation.tsx:
```diff
@@ -1,10 +1,14 @@
 
 import { SUPPORT_LANGUAGES } from '../../../common/translation/TranslationConstants';
-import replaceLocaleInPathname from '../../../common/route/utils/replaceLocaleInPathname';
+// import replaceLocaleInPathname from '../../../common/route/utils/replaceLocaleInPathname';
```
needless commented line

melniiv replied:
removed

nikomakela reviewed (commented):
There is stil lsomething wrong with menu item active indicator and footer link.

melniiv commented:
fixed active link indicator and made temporary fix for footer links. Footer links implementation is still old (which we refactored in kultus ui and admin), so it should be replaced with footer menu instead, with separate ticket. But now at least the language switch redirect works for footer links as well.

karisal-anders commented:
# Header menu testing using production CMS

## Leftmost header menu item

Works completely as it should

### Finnish
http://localhost:3000/fi/kummitoimijat/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/24074959-c323-492e-ad18-31ea35f783e3)

### Swedish
http://localhost:3000/sv/sv-kenelle/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/1c9cc084-90b9-4779-9954-549b45967a4d)

### English
http://localhost:3000/en/sponsors/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/89d47483-aee3-4534-8988-d650d18f1887)

## Middle header menu item
- Works okayish
- The **text** "Siirryt katsomaan videota Youtubeen." remains in **Finnish**
- The **slugs** for **Swedish** and **English** are **identical**

### Finnish
- http://localhost:3000/fi/tutustu-toimintaan/
- ![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/07d849b4-160f-4c60-bef4-53aedc3dc3b1)

### Swedish
- http://localhost:3000/sv/dhaqanka-iyo-mataaneynta-caruurta/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/730fd772-c34a-4e7d-9f25-0d4fa9152a3f)

### English
http://localhost:3000/en/dhaqanka-iyo-mataaneynta-caruurta/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/1d488bda-91b7-4af8-b0a0-81b264af5f71)

## Rightmost header menu item

- **Language changing doesn't work** at all, trying it will **always** go to the **root** page
- The text "Siirryt katsomaan videota Youtubeen." is always in Finnish
- The slugs for all languages are identical

### Finnish
http://localhost:3000/fi/%d8%a3%d8%b7%d9%81%d8%a7%d9%84-%d8%a7%d9%84%d8%ab%d9%82%d8%a7%d9%81%d8%a9/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/50fd6297-35ef-4f26-b1a4-06bf3d863ad4)

### Swedish
http://localhost:3000/sv/%d8%a3%d8%b7%d9%81%d8%a7%d9%84-%d8%a7%d9%84%d8%ab%d9%82%d8%a7%d9%81%d8%a9/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/a04ff775-9223-46f2-b0f1-dae612808d93)

### English
http://localhost:3000/en/%d8%a3%d8%b7%d9%81%d8%a7%d9%84-%d8%a7%d9%84%d8%ab%d9%82%d8%a7%d9%81%d8%a9/
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/f83de51a-a3f1-4577-b701-57cd71dcd14b)

karisal-anders commented:
# Footer menu testing using production CMS

## Accessibility Statement

- **Works completely** as it should, although the slugs are identical but still, it works even though they are

### Finnish
http://localhost:3000/fi/accessibility
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/f7f6e089-1418-4853-9158-4216491e6a3f)

### Swedish
http://localhost:3000/sv/accessibility
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/de95b75e-faf7-494e-b701-2d6e5377c389)

### English
http://localhost:3000/en/accessibility
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/a7a071f6-7c62-4f5d-ba13-f41148b5215d)

## Cookie settings

- **Works completely** as it should, although the slugs are identical but still, it works even though they are

### Finnish
http://localhost:3000/fi/cookie-consent
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/64746128-c548-4a74-b2c6-a1e1830bfcd5)

### Swedish
http://localhost:3000/sv/cookie-consent
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/028d77ad-ffa1-40e5-8a16-af797b19d834)

### English
http://localhost:3000/en/cookie-consent
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/68e2d7db-54f4-401b-9a91-02a414389ddc)

## Terms of Service

- Language switching changes pages but the **terms of service** are **always** in **Finnish**

### Finnish
http://localhost:3000/fi/terms
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/08b07398-c464-4151-9371-e8b76ae32123)

### Swedish
http://localhost:3000/sv/terms
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/408bd0c7-8b56-4b98-8f5e-641285f880f3)

### English
http://localhost:3000/en/terms
![image](https://github.com/City-of-Helsinki/kukkuu-ui/assets/77663720/8f503813-2cee-40ee-b9da-5e44fa03d7cd)

karisal-anders commented:
To summarize the [header](https://github.com/City-of-Helsinki/kukkuu-ui/pull/564#issuecomment-2143804854) & [footer](https://github.com/City-of-Helsinki/kukkuu-ui/pull/564#issuecomment-2143807609) testing findings:

- Header
  - Leftmost item works completely i.e. `Kulttuurikummit`
  - Text remains always in Finnish when switching languages for the middle item `Dhaqanka iyo mataaneynta caruurta` (It is only "Siirryt katsomaan videota Youtubeen." so not a huge deal)
  - **Language switching doesn't work with** the rightmost item **`أطفال الثقافة`** at all, always to root
- Footer works ok otherwise but **terms of service** are **always** in **Finnish**

karisal-anders commented on src/domain/app/navigation/Navigation.tsx:
```diff
@@ -17,27 +25,58 @@ function Navigation() {
+      });
+      const strippedPathname = stripLocaleFromUri(location.pathname);
+      // special logic for static urls
+      if (Boolean(strippedPathname)) {
```
```suggestion
      if (strippedPathname) {
```
The explicit conversion with Boolean is not needed here, or is there something here I'm overlooking?

melniiv replied:
forgot to remove, fixed

karisal-anders commented on src/domain/app/navigation/Navigation.tsx:
```diff
@@ -17,27 +25,58 @@ function Navigation() {
+          return menuItem.slug?.startsWith(strippedPathname);
+        });
+        if (staticUrl) {
+          return `/${language.toLowerCase()}${staticUrl.slug || ''}`;
```
```suggestion
          return `/${language.toLowerCase()}${staticUrl.slug ?? ''}`;
```
Sonarcloud might give some code smell/warning/note about preferring `x ?? y` instead of `x || y` so that's why I'm suggesting the use of `??` here instead of `||`

karisal-anders commented on src/domain/app/useStaticLinks.tsx:
```diff
@@ -0,0 +1,43 @@
+import useGetPathname from '../../common/route/utils/useGetPathname';
+
+// todo: this hook is used for static links in the footer navigation
+// links should be reusable so the header language switch knows how to switch the language accordinly
+// if possible, replace with header menu items from headless cms
```
```suggestion
// links should be reusable so the header language switch knows how to switch the language accordingly
// if possible, replace with footer menu items from headless cms
```
typo: accordinly -> accordingly
"replace with header menu items" -> "replace with footer menu items", right, or was there a reason for using `header` here that I'm overlooking?

karisal-anders commented on src/hooks/useCmsLanguageOptions.ts:
```diff
@@ -0,0 +1,26 @@
+  const location = useLocation();
+  const { data: pageData } = usePageQuery(location.pathname, { skip });
+
+  return !skip
+    ? [
+        {
+          uri: pageData?.page?.uri,
+          locale: pageData?.page?.language?.code?.toLowerCase(),
+        },
+        ...(pageData?.page?.translations?.map((translation) => ({
+          uri: translation?.uri,
+          locale: translation?.language?.code?.toLowerCase(),
+        })) ?? []),
+      ]
+    : [];
+};
```
This looks quite bewildering, just the nesting of different kinds of braces, ternary operator, anonymous functions etc. Looks like this code or something similar is already used in palvelutarjotin-ui and palvelutarjotin-admin... but still, looks quite bewildering, confusing. Could be refactored to add readability later.

melniiv replied:
yes, its the same, lets refactor in all projects in other Prs

karisal-anders reviewed (approved):
Makes the situation when compared to the production environment https://kummilapset.hel.fi/ better in all other cases AFAIK except the handling of the third header link (i.e. `أطفال الثقافة`) which now always go to the root page but in current production it changes between the Finnish page all the time.

Left some comments, please take a look at them. Thank you!

terovirtanen commented:
**KUKKUU-UI branch is deployed to platta: https://kukkuu-ui-pr564.dev.hel.ninja** :rocket::rocket::rocket:

karisal-anders commented on src/utils/cmsUtils.ts:
```diff
@@ -0,0 +1,11 @@
+};
+
+export const normalizeCmsUri = (uri: string) => {
+  return removeSurroundingSlashes(stripLocaleFromUri(uri));
```
```suggestion
    return decodeURIComponent(removeSurroundingSlashes(stripLocaleFromUri(uri)));
```
Fixes the changing of languages with the Arabic (i.e. the rightmost header menu item in the production headless CMS).

terovirtanen commented:
**KUKKUU-UI branch is deployed to platta: https://kukkuu-ui-pr564.dev.hel.ninja** :rocket::rocket::rocket:
